framework/include/cppmicroservices/BundleTracker.h

```diff
@@ -0,0 +1,266 @@
+
+namespace detail {
+template<class T>
+class TrackedBundle;
+template<class T>
+class BundleTrackerPrivate;
+} // namespace detail
+
+class BundleContext;
+
+/**
+\defgroup gr_bundletracker BundleTracker
+
+\brief Groups BundleTracker related symbols.
+*/
+
+/**
+ * \ingroup MicroServices
+ * \ingroup gr_bundletracker
+ * 
+ * The <code>BundleTracker</code> class allows for keeping an accurate record of bundles and handling
+ * state changes, including bundles started before the tracker is opened.
+ *
+ * <p>
+ * A <code>BundleTracker</code> is constructed with a state mask and a <code>BundleTrackerCustomizer</code> object.
+ * The <code>BundleTracker</code> can select certain bundles to be tracked and trigger custom callbacks
+ * based on the provided <code>BundleTrackerCustomizer</code> or a <code>BundleTracker</code> subclass. Additionally,
+ * the template parameter allows for the use of custom tracked objects. Once the <code>BundleTracker</code>
+ * is opened, it will begin tracking all bundles that fall within the state mask.
+ *
+ * <p>
+ * Use the <code>getBundles</code> method to get the tracked <code>Bundle</code> objects, and <code>getObject</code> 
+ * to get the customized object.
+ *
+ * <p>
+ * The <code>BundleTracker</code> class is thread-safe. It does not call <code>BundleTrackerCustomizer</code> methods while
+ * holding any locks. Customizer implementations must be thread-safe.
+ *
+ * @tparam T The type of tracked object.
+ * @remarks This class is thread safe.
+ */
+template<class T = Bundle>
+class BundleTracker : protected BundleTrackerCustomizer<T>
+{
+public:
+  // The type of tracked object
+  using TrackedParamType =
+    typename BundleTrackerCustomizer<T>::TrackedParamType;
+
+  // The type of the tracking map
+  using TrackingMap = typename std::unordered_map<Bundle, TrackedParamType>;
+
+  // The type of the state mask
+  using BundleState = std::underlying_type_t<Bundle::State>;
+
+  /**
+   * Create a <code>StateType</code> stateMask for a BundleTracker
+   * 
+   * @param s0 the <code>Bundle::State</code>.
+   * @param s the subsequent <code>Bundle::States5</code>.
+   * 
+   * @return The state mask.
+   * 
+   */
+  template <typename S0, typename... S>
+  static constexpr typename BundleTracker<T>::BundleState CreateStateMask(S0 const &s0, S const &...s);
+
+  /**
+   * Create a <code>BundleTracker</code> that tracks bundles through states covered by the state mask.
+   * 
+   * @param context The <code>BundleTrackerContext</code> from which tracking occurs.
+   * @param stateMask The bit mask which defines the bundle states to be tracked.
+   * @param customizer The customizer to call when bundles are added, modified, or removed.
+   *                   If the customizer is nullptr, then the callbacks in this <code>BundleTracker</code> will 
+   *                   be used instead (default or can be overridden).
+   */
+  BundleTracker(
+    const BundleContext& context,
+    BundleState stateMask,
```

Comment:
Wondering if the stateMask can be a const. Does it get changed after it's initialized?